### PR TITLE
Scrub passwords from error output

### PR DESCRIPTION
**Pull request by mkenney**
This is in reference to https://github.com/go-goracle/goracle/issues/79.

### CODE REVIEW

**Comment by mkenney** [comment]
There's an s3 credentials error in the build trying to fetch the instantclient.

**Comment by mkenney** [comment]
@tgulacsi, `ConnectionParams` does have a [`String`](https://github.com/ReturnPath/goracle-pr/blob/master/drv.go#L521) method which is what's being used for the error output (which is the root of the issue).

I agree the two things should be separate, and I don't particularly like intercepting and sanitizing the flow this way, but it also smells bad to do to much more around scoping here as suggested by @veqryn without being more familiar with the consequences for downstream projects.

My only goal is to prevent unexpected or unintended output of sensitive data. I thought about maybe adding an `error` or even an `Error` method. `Error` feels weird, making `ConnectionParams` an error, but an unexported `error` method might make more sense. Thoughts?

**Comment by tgulacsi** [comment]
I wanted to say that changing line 535 of drv.go from

    User:   url.UserPassword(P.Username, P.Password),

to

    User:   url.UserPassword(P.Username, "***"),

would be enough, wouldn't it?

**Comment by mkenney** [comment]
Ahh, I see. Ok, I'll take a closer look and push up a change. Thanks!

**Comment by mkenney** [comment]
So it looks like `ConnectionParams.string` is used for both logging and creating the actual connection string so changing 535 to `User:   url.UserPassword(P.Username, "***"),` will break some things unless I'm missing something.

I'll find those cases (there might just be the one) and split them out to use another method or maybe add a param to the `string` method...

**Comment by tgulacsi** [comment]
Where is ConnectionParam.String (or string) used for the connection?

**Comment by mkenney** [comment]
It's in [`openConn`](https://github.com/go-goracle/goracle/blob/master/drv.go#L367).

I pushed up a change that I think will cover all the bases without breaking anything. I chose `SECRET` as the password replacement string instead of asterisks because `url.UserPassword` url encodes the string making the asterisks `%2A` which feels odd. 

Please let me know what you think.

File: drv.go [hunk 1]
```diff
@@ -446,7 +444,7 @@ func (d *drv) openConn(P ConnectionParams) (*conn, error) {
 	if P.IsSysDBA || P.IsSysOper {
 		dc := C.malloc(C.sizeof_void)
 		if Log != nil {
-			Log("C", "dpiConn_create", "username", P.Username, "password", P.Password, "sid", P.SID, "common", commonCreateParams, "conn", connCreateParams)
+			Log("C", "dpiConn_create", "username", P.Username, "password", "SECRET", "sid", P.SID, "common", commonCreateParams, "conn", connCreateParams)
```
Comment (tgulacsi) [inline drv.go]: Just delete the printing of "password", and its value - this provides no information.

File: drv.go [hunk 2]
```diff
@@ -457,7 +455,7 @@ func (d *drv) openConn(P ConnectionParams) (*conn, error) {
 			&connCreateParams,
 			(**C.dpiConn)(unsafe.Pointer(&dc)),
 		) == C.DPI_FAILURE {
-			return nil, errors.Wrapf(d.getError(), "username=%q password=%q sid=%q params=%+v", P.Username, P.Password, P.SID, connCreateParams)
+			return nil, errors.Wrapf(d.getError(), "username=%q password=%q sid=%q params=%+v", P.Username, "SECRET", P.SID, connCreateParams)
```
Comment (tgulacsi) [inline drv.go]: Just leave it out.

File: drv.go [hunk 3]
```diff
@@ -480,7 +478,7 @@ func (d *drv) openConn(P ConnectionParams) (*conn, error) {
 
 	var dp *C.dpiPool
 	if Log != nil {
-		Log("C", "dpiPool_create", "username", P.Username, "password", P.Password, "sid", P.SID, "common", commonCreateParams, "pool", poolCreateParams)
+		Log("C", "dpiPool_create", "username", P.Username, "password", "SECRET", "sid", P.SID, "common", commonCreateParams, "pool", poolCreateParams)
```
Comment (tgulacsi) [inline drv.go]: Just leave it out.

File: drv.go [hunk 4]
```diff
@@ -492,7 +490,7 @@ func (d *drv) openConn(P ConnectionParams) (*conn, error) {
 		(**C.dpiPool)(unsafe.Pointer(&dp)),
 	) == C.DPI_FAILURE {
 		return nil, errors.Wrapf(d.getError(), "username=%q password=%q SID=%q minSessions=%d maxSessions=%d poolIncrement=%d extAuth=%d ",
-			P.Username, strings.Repeat("*", len(P.Password)), P.SID,
+			P.Username, "SECRET", P.SID,
```
Comment (tgulacsi) [inline drv.go]: Just leave it out.

File: drv.go [hunk 5]
```diff
@@ -364,12 +364,12 @@ func (d *drv) openConn(P ConnectionParams) (*conn, error) {
 	}
 
 	c := conn{drv: d, connParams: P}
-	connString := P.StringNoClass()
+	connString := P.string(false, true)
```
Comment (tgulacsi) [inline drv.go]: We don't need a new pool per-password, so we don't need the password in the connString, so we don't need the second parameter in P.string.

Comment (mkenney) [reply drv.go]: Ok, in that case I don't think the second param is needed at all and `string` can always leave out the password. I'll push an update shortly.

**Comment by mkenney** [comment]
Ok, I made those updates and connections look good in my test script. Let me know what you think, thanks!

**Comment by mkenney** [comment]
Awesome, thank you for your help!

**Pull request merged**